### PR TITLE
WIP: Graph smoothing

### PR DESCRIPTION
**Pull request by benleetownsend**
Accidentally applied black to target_blocks.py but the important diff for this is between 434-446

### CODE REVIEW

File: finetune/nn/graph_conv.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+    def call(self, inputs):
+        features, adjacency_mat = inputs
+        features_with_head_dim = tf.expand_dims(features, 1)  # batch, 1, seq, feats
+        # hopefully this will broadcast across heads dim?
+        projected = tf.matmul(features_with_head_dim, self.kernel)
+        # projected = batch, heads, seq, n_out
+        # adjacency_mat = batch, heads, seq, seq
+        adjacency_mat = tf.nn.relu(adjacency_mat)  # remove any spurious negative values
+        normed_adjacency_mat = tf.math.divide_no_nan(
```
Comment (benleetownsend) [inline finetune/nn/graph_conv.py]: There are much more well reasoned normalisation approaches that can be applied to the adjacency matrix

File: finetune/nn/graph_conv.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+    low_values_y = tf.expand_dims(values[low_key], 2)  # batch, seq, 1
+    high_values_y = tf.expand_dims(values[high_key], 2)  # batch, seq, 1
+    return {
+        "before": tf.cast(
+            high_values_y < low_values_x, tf.float32
+        ),  # for dims (batch, in, out) this says in is above out,
+        "after": tf.cast(
+            low_values_y > high_values_x, tf.float32
+        ),  # for dims (batch, in, out) this says in is above out,
```
Comment (benleetownsend) [inline finetune/nn/graph_conv.py:38]: is below out

File: finetune/nn/graph_conv.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        # projected = batch, heads, seq, n_out
+        # adjacency_mat = batch, heads, seq, seq
+        adjacency_mat = tf.nn.relu(adjacency_mat)  # remove any spurious negative values
+        normed_adjacency_mat = tf.math.divide_no_nan(
+            adjacency_mat, tf.reduce_sum(adjacency_mat, 2, keepdims=True)
+        )
+        shared = tf.matmul(
+            projected, normed_adjacency_mat, transpose_a=True
+        )  # batch, n_out, heads, seq
+        identity_out = tf.matmul(features, self.identity_kernel)
```
Comment (benleetownsend) [inline finetune/nn/graph_conv.py:192]: The graph where everything is connected only to its self

File: tests/test_utils.py [hunk 4]
```diff
@@ -74,6 +74,72 @@ class TestRobertaV2Encoder(TestGPTEncoder):
...
+            이것은 되려니와, 이성은 노년에게서 있는 밥을 거친 봄바람이다. 
+            있는가? 기관과 없으면. 봄바람이다. 
+            날카로우나 없는 그러므로 봄바람이다.
+            """
+        )
+
+    def test_token_alignment_japanese_chinese_english_mix(self):
+        self.tokenize_and_assert_equlivalence( 
+            """
+            Chinese:
```
Comment (lhz1029) [inline tests/test_utils.py]: I think you have the Chinese and Japanese flipped; shouldn't affect anything, just something to note

File: finetune/nn/graph_conv.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        )
+
+    def call(self, inputs):
+        features, adjacency_mat = inputs
+        features_with_head_dim = tf.expand_dims(features, 1)  # batch, 1, seq, feats
+        # hopefully this will broadcast across heads dim?
+        projected = tf.matmul(features_with_head_dim, self.kernel)
+        # projected = batch, heads, seq, n_out
+        # adjacency_mat = batch, heads, seq, seq
+        adjacency_mat = tf.nn.relu(adjacency_mat)  # remove any spurious negative values
```
Comment (lhz1029) [inline finetune/nn/graph_conv.py:182]: wouldn't it be a bug if any existed? maybe it's worth adding a warning for that case?

File: finetune/nn/graph_conv.py (hunk 3 above)
Comment (lhz1029) [reply finetune/nn/graph_conv.py:192]: I'm not sure I understand what you mean here? It seems like identity_kernel and kernel are both just learned weights, with the latter getting an additional smoothing applied based on the adjacency matrix afterwards?

File: finetune/nn/graph_conv.py (hunk 1 above)
Comment (lhz1029) [reply finetune/nn/graph_conv.py]: Part 2 mentions some normalization techniques: http://tkipf.github.io/graph-convolutional-networks/

File: finetune/nn/graph_conv.py (hunk 3 above)
Comment (benleetownsend) [reply finetune/nn/graph_conv.py:192]: So the adjacency matrices explicitly do not connect nodes to themselves. This could be solved by adding an extra adjacency matrix that is just the identity. But this seems cleaner.

Comment (lhz1029) [reply finetune/nn/graph_conv.py:192]: To achieve that, would it make sense to be adding projected and shared? identity_out seems to be different, since it's derived from a completely different set of weights applied to features. Is there a reason for that in particular?

Comment (benleetownsend) [reply finetune/nn/graph_conv.py:192]: So each of the graphs has it's own independent sets of weights as well they just happen to be in the same tensor.

Comment (benleetownsend) [reply finetune/nn/graph_conv.py:192]: So unless I've messed something up this should be exactly equivalent to having an additional graph that is the identity matrix

Comment (lhz1029) [reply finetune/nn/graph_conv.py:192]: ah got it. I thought you wanted to add nodes to themselves in the already existing graphs, but this also seems reasonable

File: finetune/nn/graph_conv.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+
+def _distance(values, low_key, high_key):
+    middle_values = (values[low_key] + values[high_key]) // 2
+    return tf.abs(tf.expand_dims(middle_values, 1) - tf.expand_dims(middle_values, 2))
+
+
+def similarity(values):
+    return {
```
Comment (madisonmay) [inline finetune/nn/graph_conv.py:59]: This function is rather sharp near 0 dist when used w/ L1 distance?

File: finetune/nn/graph_conv.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        * similarity_matrix["x"]
+        * mask,
+        "right": direction_matrix["right"]
+        * overlap_matrix["y"]
+        * similarity_matrix["x"]
+        * mask,
+    }
+
+
+class PairwiseDistance(tf.keras.layers.Layer):
```
Comment (madisonmay) [inline finetune/nn/graph_conv.py:99]: Currently unused seems like?

File: finetune/nn/graph_conv.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+        projected = tf.matmul(features_with_head_dim, self.kernel)
+        # projected = batch, heads, seq, n_out
+        # adjacency_mat = batch, heads, seq, seq
+        adjacency_mat = tf.nn.relu(adjacency_mat)  # remove any spurious negative values
+        normed_adjacency_mat = tf.math.divide_no_nan(
+            adjacency_mat, tf.reduce_sum(adjacency_mat, 2, keepdims=True)
+        )
+        shared = tf.matmul(
+            projected, normed_adjacency_mat, transpose_a=True
+        )  # batch, n_out, heads, seq
```
Comment (madisonmay) [inline finetune/nn/graph_conv.py]: Think this is `(batch, heads, n_out, seq)`
